src/compiler/ir/README.md

```diff
@@ -0,0 +1,162 @@
+Optimizations may enable us to dispatch the call statically, using for example,
+the type of the receiver.
+
+The method being called could be static or not (ie. does it accept a `self`
+parameter, not to be confused with the dispatch strategy). In either case, the
+receiver of the call is always passed as the first argument to the method.  In
+the case of a static method call, the receiver may not be an actual object, but
+only a pointer to a descriptor, as created by a `StaticTypeStmt`.
+
+## `WhenStmt`
+
+Schedule a behaviour on a set of cown. The `closure_index` corresponds to the
+index of the body within the `MethodIR`.
+
+This corresponds to a `rt::Cown::schedule` call. The message sent to the cowns
+should provide storage for a pointer to the closure's body, the list of cowns
+and the captures.
+
+The closure's body will accept as many parameters as there are cowns and
+captures specified. When invoking the body, each cown argument should be
+transformed into a reference to its contents, before being passed as an
+argument. The captures should be passed as is.
+
+## `ReadFieldStmt`
+
+Read a field from an object. The field is identified by its name. The general
+case is similar to method calls: the field's selector is used to look up the
+field's offset in the object's vtable.
+
+## `WriteFieldStmt`
+
+Write a field into an object. The procedure for locating the offset of the
+field is identical to a field read.
+
+When writing a reference counted value to an object in a region that is
+equipped with a RememberedSet, the value should be inserted into the RS.
+Otherwise, the reference count should be incremented.
+
+## `IntegerLiteralStmt`, `StringLiteralStmt`, `UnitStmt`
+
+These create values from a literal. The support for strings is very special
+cased for demonstration purposes.
+
+## `MatchBindStmt`
+
+This is an unchecked cast, used at the top of the target of a match statement.
+
+## `ViewStmt`
+
+If the argument to the statement is an owned value, the statement produces an
+unowned reference to the same object. Otherwise this is equivalent to a copy.
+
+## `EndScopeStmt`
+
+This statement is inserted whenever variables go out of scope. This may not be
+the end of a BasicBlock, if a block expression is used for example.
+
+This is where a value's "release" code should be executed. This could be
+`rt::Region::release`, `rt::Cown::release` or `rt::Immutable::release`, for
+owned, cown and immutable values, respectively. Unowned values do not need to
+be released.
+
+## `OverwriteStmt`
+
+This is used after a local variable assignement, on the old SSA value bound to
+this statement. The old value should be released, as in an `EndScopeStmt`. For
+example, the code `var x = new A; x = new B; ...`, would be lowered to the
+following IR:
+
+```
+0 <- new A
+1 <- new B
+overwrite(0)
+// ...
+end-scope(1)
+```
+
+The local `x` is originally assigned the SSA variable `0`, and later `1`. After
```

Comment:
Why isn't this just `end-scope(0)`? If the variable `x` was updated in the symbol table to point to `1`, then `0` is dangling and `end-scope` would just free it.

Reply:
Meh, `end-scope` and `overwrite` basically have the same semantics, but we want different error messages out of them.

Reply:
Fair enough.